File: c/src/sort/count_sort.c
#include <stdlib.h>

#include "sort.h"
#include "utils.h"

void count_sort(int* arr, size_t size) {
  int k = array_max_value(arr, size) + 1;
  int* counter = (int*)calloc(k,sizeof(int));
  for (size_t i = 0; i < size; ++i) {
    counter[arr[i]]++;
  }
  int i = 0;
  int j = 0;
  while (j < k) {
    if (counter[j] != 0) {
      arr[i++] = j;
      --counter[j];
    } else {
      ++j;
    }
  }
  free(counter);
}

unsigned long long count_sort_book_ram(int* arr, size_t size) {
  return 0;
}

unsigned long long count_sort_ram(int* arr, size_t size) { return 0; }

File: rust/src/algorithms/sort/count_sort.rs
fn array_max_value(arr: &[usize]) -> usize {
    let mut max = arr[0];
    for &x in arr.iter() {
        if x > max {
            max = x;
        }
    }
    max
}

pub fn count_sort(arr: &mut [usize]) {
    let k = array_max_value(arr) + 1;
    let mut counter = vec![0usize; k];
    for &x in arr.iter() {
        counter[x] += 1;
    }
    let mut i = 0usize;
    let mut j = 0usize;
    while j < k {
        if counter[j] != 0 {
            arr[i] = j;
            i += 1;
            counter[j] -= 1;
        } else {
            j += 1;
        }
    }
}
